pkg/controller/controller.go

```diff
@@ -0,0 +1,77 @@
+package controller
+
+import (
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	kscheme "k8s.io/client-go/kubernetes/scheme"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/operator-framework/combo/api/v1alpha1"
+)
+
+var (
+	schemeBuilder = runtime.NewSchemeBuilder(
+		kscheme.AddToScheme,
+		v1alpha1.AddToScheme,
+	)
+
+	// AddToScheme adds all types necessary for the controller to operate.
+	addToScheme = schemeBuilder.AddToScheme
+)
+
+type manageable interface {
+	manageWith(mgr ctrl.Manager) error
+}
+
+type manageables []manageable
+
+func (m manageables) manageWith(mgr ctrl.Manager) error {
+	for _, man := range m {
+		if m == nil {
+			// Panic because this case indicates a bug
+			panic("failed to register controller with manager: cannot be nil")
+		}
+
+		if err := man.manageWith(mgr); err != nil {
+			// Bail out if any controller fails to register with the manager
+			return err
+		}
+	}
+
+	return nil
+}
+
+type Controller struct {
+	client.Client
+	log     logr.Logger
+	managed manageables
+}
+
+// NewReconciler constructs and returns a controller.
+func NewController(cli client.Client, log logr.Logger) (*Controller, error) {
+	return &Controller{
+		Client: cli,
+		log:    log,
+		managed: manageables{
+			&templateController{
+				Client: cli,
+				log:    log.WithValues("controller", "template"),
+			},
+			&combinationController{
+				Client: cli,
+				log:    log.WithValues("controller", "combination"),
+			},
```

Comment:
I find this kind of structure a bit awkward to develop on personally, especially when the controller is expecting extra configuration (e.g. uses the options pattern) that cannot be handled by defaulting values, but this is fine for now.

Reply:
I could see this being refactored to utilize a factory function. Let me take a stab at that.

Reply:
This could be an alternative that I came up with to get around the no logical default issue:

```go
managed: manageables{
    template.NewController(cli, log.WithValues("contoller", "template")),
    combo.NewController(cli, log.WithValues("contoller", "combination")),
},
```

Reply:
That looks reasonable but I think we just hold off on refactoring for now and if the implementation is too cumbersome when trying to configure multiple controllers, we can take a stab then.